### PR TITLE
Spark Notebook in viewer mode (only) 

### PR DESCRIPTION
**Pull request by andypetrella**
This is necessary to deploy notebook in presentation/read-only mode in any website without opening the doors to hell (for the website manager)... which is REPL :-).

This is separating spark from the main project if it is build if the viewer mode enabled (sbt `-Dviewer=true`).

Hence the spark module is then not included in the binary (however, spark are still there due to dependencies in the widgets).

This gets us closer to separate more the server from spark kernel with specific version for instance. But this wasn't the main goal here.

Downside, the `Repl` is now created using reflection... because the ReplCalculator cannot be extracted since it is used in Application.scala (cf separation remark above).

Leftover, the actions like create notebook and so forth are only disabled client side, this should be addressed on the server side too.
However, the most harmful thing, which is the opening the `REPL` isn't possible at all!

cc @vidma @meh-ninja @virtualirfan

### CODE REVIEW

File: public/ipython/tree/js/notebooklist.js [hunk 1]
```diff
@@ -274,6 +275,14 @@ define([
     };
 
 
+    NotebookList.prototype.read_only_url = function(url) {
```
Comment (vidma) [inline public/ipython/tree/js/notebooklist.js]: `read_only_url` should be stateless.

**Comment by vidma** [comment]
I don't know...
what's the tangible use of REPL-less viewer mode (and even more, enabled a compile time, in SBT)? is it worth the extra complexity?
If it was URL parameter I would support this more..

**Comment by andypetrella** [comment]
Let me elaborate a bit maybe. The need here is to have a viewer that we can deploy on any website without opening the doors to the machine.

There are several use cases, where the simpler is to have demo website, or tutorial, or even simply documentation (for an algorithm).

The extra complexity (specially in SBT) might not be needed but interesting to have, if we want to separate the kernel/repl from the server, which would be interesting if we want, for instance, a REPL for scala only.

Put the SBT thingy aside, we can refactor how this is working by having this mode enabled from the configuration for the whole server (I mean, don't use BuildInfo for that).

File: public/ipython/tree/js/notebooklist.js (hunk 1 above)
Comment (andypetrella) [inline public/ipython/tree/js/notebooklist.js]: Since it is attached to the notebook list instance, you don't think it's fair?

Or you propose to have a static function that would take both `url` and `viewer` as params?

**Comment by andypetrella** [comment]
Looks like the scala 2.11 branch hasn't been refactored to use a common REPL trait, so I'll do it too.
Then we can remove a lot of things from the specific modules.

This will also allow other kind of REPLs, since the API will be dictated by it's usage in the ReplCalculator (which is still very coupled with Spark though...)

File: modules/kernel/src/main/scala-2.10/notebook/ReplCalculator.scala [hunk 2]
```diff
@@ -230,7 +230,7 @@ class ReplCalculator(
           log.info(s"Killing job Group $jobGroupId")
           val thisSender = sender()
           repl.evaluate(
-            s"""sparkContext.cancelJobGroup("${jobGroupId}")""",
+            s"""globalScope.sparkContext.cancelJobGroup("${jobGroupId}")""",
```
Comment (vidma) [inline modules/kernel/src/main/scala-2.10/notebook/ReplCalculator.scala]: does it work ?
I remember there was some difference between 2.10 & 2.11...

Comment (andypetrella) [inline modules/kernel/src/main/scala-2.10/notebook/ReplCalculator.scala]: yes, serializable crap in 2.11 I think.

Indeed we can put things in different PRs too...

**Comment by vidma** [comment]
this is very old, I think.
and few changes see unrelated...
